Setup the project

IbrahimMurad opened the pull request:
This pull request includes the initial setup and configuration for the Carhub Django project. The most important changes include adding essential configuration files, setting up the Django project structure, and configuring development tools.

### Development Tools Configuration:

* [`.flake8`](diffhunk://#diff-6951dbb399883798a226c1fb496fdb4183b1ab48865e75eddecf6ceb6cf46442R1-R5): Configured flake8 for linting with specific settings for ignored files, maximum line length, and complexity.
* [`pyproject.toml`](diffhunk://#diff-50c86b7ed8ac2cf95bd48334961bf0530cdc77b5a56f852c5c61b89d735fd711R1-R48): Added project metadata, dependencies, and configuration for development tools like black and isort.
* [`vscode/settings.json`](diffhunk://#diff-75113530365f5f42f8a09821bec608e4a45936a14d82caaef865fb7350cbb010R1-R18): Configured VS Code settings for Python development, including formatting, linting, and testing.

### Environment Configuration:

* [`.env.dist`](diffhunk://#diff-84f0ebf34c2eaf5b3b8f8263f855938cef61b0fa67cd0528279daa28ef40a096R1-R2): Added an example environment file with a placeholder for the secret key.

### Dependencies:

* [`requirements`](diffhunk://#diff-10762e1094dbeeaba59c58cec8a91b71b99f2063f7f1f20e5c0e3e635f42b2dfR1-R2): Added dependencies for Django and django-environ.

### Starting the Project

IbrahimMurad committed:
Add configuration files for Flake8, Git, and project setup with pyproject.toml

IbrahimMurad committed:
Add Django and development dependencies to pyproject.toml and create requirements file

IbrahimMurad committed:
start django project, installing django-environ and including it in requirements and pyproject dependencies, setting the secret key in .env file

IbrahimMurad committed:
Add user management features and database configuration

- Updated .env.dist to include database configuration options.
- Modified settings.py to use environment variables for database settings.
- Created User model with custom role choices and phone number field.
- Added AdminsManager, CustomersManager, and SellersManager for user role management.
- Registered User model in admin.py.
- Created initial migration for User model.
- Added base model for future models in core/models.py.
- Updated .gitignore to exclude __pycache__.

hassaanalansary commented on .flake8:
```diff
@@ -0,0 +1,5 @@
+[flake8]
+per_file_ignores = __init__.py:F401, settings.py:E501
```
I think you can add all the flake8 config in pyproject.toml

IbrahimMurad replied:
I tried to do it before (including it in pyproject.toml) but it either does not work or throws many errors.

hassaanalansary commented on Carhub/asgi.py:
```diff
@@ -0,0 +1,16 @@
+
+from django.core.asgi import get_asgi_application
+
+os.environ.setdefault('DJANGO_SETTINGS_MODULE', 'Carhub.settings')
```
The apps and main app should be lower case

hassaanalansary commented on requirements:
```diff
@@ -0,0 +1,2 @@
+django
```
You should use a specific version, so other developera will have the same experience 
Also you can add requirements_dev.text that contain requirements related to development and testing
Such as black, pre-commit, pytest, ....

hassaanalansary commented on users/models.py:
```diff
@@ -0,0 +1,74 @@
+#         return self.address + ", " + self.city + ", " + self.Country
+
+
+class AdminsManager(models.Manager):
```
I prefer it should be singular 
```suggestion
class AdminManager(models.Manager):
```

hassaanalansary commented on users/models.py:
```diff
@@ -0,0 +1,74 @@
+
+    objects = models.Manager()
+    admins = AdminsManager()
+    customers = CustomersManager()
```
Having multiple manags like this is acceptable but I am not a big fan, I believe this could be done with a queryset or on the fly

hassaanalansary commented on users/models.py:
```diff
@@ -0,0 +1,74 @@
+        """choices for the role field."""
+
+        ADMIN = "admin", "Admin"
+        Customer = "Customer", "Customer"
```
```suggestion
        CUSTOMER = "Customer", "Customer"
```

hassaanalansary commented on users/models.py:
```diff
@@ -0,0 +1,74 @@
+        SELLER = "seller", "Seller"
+
+    phone_number = models.CharField(max_length=15, blank=True, null=True)
+    role = models.CharField(
```
I believe that having a profile OneToOne relation is a better approach
Because it will seperate responsibility and it will allow the same User to have multiple roles at the same time

IbrahimMurad committed:
changing the project name to lowercase.
including the version in the requirements file and creating a requirements_dev file.
Removing the managers.
Reimplementing the user model to use a profile model that has a one-to-one relationg with the built-in user model.
